{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dramatic-blair",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "import sys\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from collections import OrderedDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc68c68b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Table that contains three columns: genotype names, FASTA and GFF3 names\n",
    "fdf = pd.read_csv(\"files_names.csv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "034363a4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ext_len = 500 # The length of 3' UTR extension if required\n",
    "for gff_row in fdf.iterrows():\n",
    "    print(gff_row[0])\n",
    "    # Load fasta file into dictionary for easy access. Can comment out when loaded for development\n",
    "    with gzip.open(\"Genome/\"+gff_row[1][\"FASTA\"], \"rt\") as handle:\n",
    "        fasta_dict = {record.id: record.seq for record in SeqIO.parse(handle, \"fasta\")}\n",
    "\n",
    "    gff = pd.read_csv(\"GFF3/\"+gff_row[1][\"GFF3\"], header=None, sep=\"\\t\", comment='#', compression='gzip', low_memory=False)\n",
    "    # gff = gff.iloc[:100] # For development work with fewer lines\n",
    "\n",
    "    gene_od = OrderedDict() # dictionary of geneID: Seq after extension\n",
    "    skip_first = True \n",
    "    skip_until_next_mRNA = True # Only dealing with canonical/primary transcripts\n",
    "    for _, row in gff.iterrows():\n",
    "        feature = row[2] # gene, mRNA, five_prime_UTR, CDS or three_prime_UTR\n",
    "\n",
    "        # Skip rows if the associated mRNA is not primary\n",
    "        if feature != \"mRNA\" and skip_until_next_mRNA:\n",
    "            continue\n",
    "\n",
    "        # Build the cDNA sequence from primary sequences. Non-primary skip in previous loop\n",
    "        if feature in [\"CDS\", \"five_prime_UTR\",\"three_prime_UTR\"]:\n",
    "                        \n",
    "            chrom = row[0]\n",
    "            start = row[3]-1\n",
    "            end = row[4]\n",
    "            strand = row[6]\n",
    "            three_prime_len = 0\n",
    "            if geneID not in gene_od.keys():\n",
    "                gene_od[geneID] = \"\"\n",
    "            if strand == \"+\":\n",
    "                gene_seq[feature] += fasta_dict[chrom][start:end]\n",
    "            else:\n",
    "                gene_seq[feature] += fasta_dict[chrom][start:end].reverse_complement()\n",
    "\n",
    "\n",
    "        if feature == \"mRNA\":\n",
    "            split = row[8].split(\";\")\n",
    "            # Check if mRNA marks the beginning of a primary transcript section\n",
    "            if \"canonical_transcript=1\" in split or \"longest=1\" in split:\n",
    "                \n",
    "                # The sequences are meant to be written when reaching a new mRNA and are \n",
    "                # not initiated when the loop first starts. Skip writing empty gene on first iter.\n",
    "                if skip_first:\n",
    "                    skip_first = False\n",
    "                else:\n",
    "                    # Write the fast3p SeqRecord to the ordered dict that will be saved to fasta\n",
    "                    tpu_len = len(gene_seq[\"three_prime_UTR\"])\n",
    "                    if tpu_len < ext_len:\n",
    "                        if strand == \"+\":\n",
    "                            gene_seq[\"three_prime_UTR\"] += fasta_dict[chrom][end:end+ext_len-tpu_len]\n",
    "                        else:\n",
    "                            gene_seq[\"three_prime_UTR\"] += fasta_dict[chrom][start-ext_len-tpu_len:start].reverse_complement()\n",
    "                    # Seq(gene_seq[\"five_prime_UTR\"] + gene_seq[\"CDS\"] + gene_seq[\"three_prime_UTR\"]) # will build whole gene\n",
    "                    gene_od[geneID] = SeqRecord(seq=Seq(gene_seq[\"CDS\"][-500:] + gene_seq[\"three_prime_UTR\"]), id=geneID, name='', description='')\n",
    "\n",
    "                \n",
    "                # Next few rows are associated with primary transcript and should not be skipped\n",
    "                skip_until_next_mRNA = False\n",
    "                \n",
    "                # Initiate a dictionary that will contain the feature sequences\n",
    "                for s in split:\n",
    "                    if \"Parent=\" in s:\n",
    "                        geneID = s[7:].replace(\".RefGen_V4\", \"\")\n",
    "                        gene_seq = {\"CDS\": \"\", \"five_prime_UTR\": \"\", \"three_prime_UTR\": \"\"}\n",
    "                        #print(geneID)\n",
    "                \n",
    "            # When there are multiple variants then skip non-primary rows\n",
    "            else:\n",
    "                skip_until_next_mRNA = True\n",
    "                \n",
    "    # At the final gene there will be no more primary genes to write the last gene\n",
    "    # Write the fast3p SeqRecord to the ordered dict that will be saved to fasta\n",
    "    tpu_len = len(gene_seq[\"three_prime_UTR\"])\n",
    "    if tpu_len < 1000:\n",
    "        if strand == \"+\":\n",
    "            gene_seq[\"three_prime_UTR\"] += fasta_dict[chrom][end:end+1000-tpu_len]\n",
    "        else:\n",
    "            gene_seq[\"three_prime_UTR\"] += fasta_dict[chrom][start-1000-tpu_len:start].reverse_complement()\n",
    "    # Seq(gene_seq[\"five_prime_UTR\"] + gene_seq[\"CDS\"] + gene_seq[\"three_prime_UTR\"]) # will build whole gene\n",
    "    gene_od[geneID] = SeqRecord(seq=Seq(gene_seq[\"CDS\"][-500:] + gene_seq[\"three_prime_UTR\"]), id=geneID, name='', description='')\n",
    "            \n",
    "    with open('fast3p/'+gff_row[0]+'_fast3p.fa', 'w') as handle:\n",
    "        SeqIO.write(gene_od.values(), handle, 'fasta')\n",
    "    \n",
    "    #break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8e230fd",
   "metadata": {},
   "source": [
    "# Attempt (failed) to adjust overlapping UTRs and CDS sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33cf19fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "limit_dict = {}\n",
    "for rx in range(len(primary_coord_list)):\n",
    "    print(primary_coord_list[rx+1][0] - primary_coord_list[rx][1])\n",
    "    if primary_coord_list[rx+1][0] - primary_coord_list[rx][1] < 2000:\n",
    "        print(\"Too close\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afbb4583",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To avoid eptending a 3' UTR into another gene we will first get the coordinates of all genes\n",
    "# Then we can index the list and check that the extension doesn't go into the next gene\n",
    "# If primary gene models are more than 2000 bp apart we can skip them for the time being\n",
    "primary_gene_list = []\n",
    "primary_coord_list = []\n",
    "geneID = \"\" \n",
    "skip_until_next_mRNA = True # Only dealing with canonical/primary transcripts\n",
    "for _, row in gff.iterrows():\n",
    "    feature = row[2] # gene, mRNA, five_prime_UTR, CDS or three_prime_UTR\n",
    "        \n",
    "    # Skip rows if the associated mRNA is not primary\n",
    "    if feature != \"mRNA\" and skip_until_next_mRNA:\n",
    "        continue\n",
    "        \n",
    "        \n",
    "    if feature == \"mRNA\":\n",
    "        split = row[8].split(\";\")\n",
    "        # Check if \n",
    "        if \"canonical_transcript=1\" in split or \"longest=1\" in split:\n",
    "            skip_until_next_mRNA = False\n",
    "            for s in split:\n",
    "                if \"Parent=\" in s:\n",
    "                    primary_gene_list.append(s[7:].replace(\".RefGen_V4\", \"\"))\n",
    "                    primary_coord_list.append([row[3]-1, row[4], row[6]]) # list of start, end, strand                    \n",
    "        else:\n",
    "            skip_until_next_mRNA = True"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
